big-neon/bn-api
# WIP: Access codes

Closes: #448

This PR attempts to address the associated issue by adding a new `CodeTypes::Access` type for use with the API. In addition several smaller changes were made in the branch.

I originally intended to look into adding the ticket type pricing and add to cart logic here but #509 will take care of most of that and it's waiting on #507 so figured it was better to just get this out as is and we can iterate on it. In addition my schedule for the next week is going to be erratic so don't want it holding things up. I'll make a note in #509 to also add the gating functionality of the `Access` type when we do the ticket types pricing logic changes.

Todo:
- Still need to spend a little more time with the branch to make sure nothing is amiss.
- Going to add pagination support to the codes endpoint (will make a new issue for the events and other endpoints currently just returning the pagination struct without doing anything on the db side to support it)
- Make a note about adding access gating to #509
- Confirm the db validations added won't cause any issues (I can't add tests at the moment so can't fully verify but will once over them again and then in #509 we can add some tests to further validate it)

Changes:
- Adds support to make `discount_in_cents` nullable adjusts validations around that
- Moved the `code_id` off of the `orders` and `ticket_instances` tables in favor of tracking via the `order_items` as the add to cart logic coming in #507 used that table to track redemption code associations with the order. I was able to pull together what I needed for the validations based on that new field (as we can just group by order and count those to check against `max_uses` as done in the validation).
- Fixed codes endpoint to take `code_type` and mimic events index setup
- Adds cascading deletions to `comps` if the associated `holds` row is deleted
- Clears Comps when Hold type is changed (there's a todo there to prevent changing the type when Comp already used so that aspect wasn't handled here)
- Cleans up the validation logic a little bit to fit a standard format across the models

Questions:
- Are we going to ever hit the codes endpoint without a code type? Or was that done just temporarily? Wondering if I can nuke the option there or if we'll have a reason to pull all the discount and access codes at once (perhaps some export? although I feel we'd want more control of the output then).
- Is this new format correct for the codes search parameters? Not sure how the tags are supposed to work so mimicked the events endpoint. Are the tags supposed to be how we transmit that data or just additional data returned about how the query worked?

Validations:
- Check quantity of order items against code max_tickets_per_user
- Check number of orders against code max uses
- Confirms discount is present for `CodeTypes::Discount` but not required (but still usable) for `CodeTypes::Access`

Notes:
- I didn't add any order item creation validation tests as there's no way to create one with a code at the moment. That will be possible with #509, however.
- The codes logic doesn't currently support the paging parameters but neither does most of the code (it seems like that feature is mostly just for showing the correct format for now? Nothing I could find seems to be hooked up to use it). Will be at least adding support for this as mentioned in the todo above.
- Tried to switch the query parameter logic to use serde flatten but it appears it fails with actix-web as it treats them as strings so going to hold off on that for now. Will file an issue with actix-web later about it once I have a chance to make a concise test case.

https://serde.rs/attr-flatten.html
```
#[serde(flatten)]
paging_parameters: PagingParameters,

curl "http://localhost:9000/events?page=2" -i
HTTP/1.1 400 Bad Request
{ "level": "WARN", "target": "actix_web::pipeline", "message": "Error occured during request handling: invalid type: string "2", expected u32" }
{ "level": "DEBUG", "target": "actix_web::pipeline", "message": "Error { err: "invalid type: string \"2\", expected u32" }" }
```
